README.md
C'est le code de référence pour le protocole crypto-monnaie [CryptoNote] (https://cryptonote.org).

* Lancez votre propre devise CryptoNote: [CryptoNote Starter] (https://cryptonotestarter.org/)
* Implémentation de référence CryptoNote: [CryptoNoteCoin] (https://cryptonote-coin.org)
* Forum de discussion et support: [Forum CryptoNote] (https://forum.cryptonote.org)

## CryptoNote comment-faire

### Preparation

1. Créez un compte sur [GitHub.com] (github.com)
2. Branch [référentiel CryptoNote] (https://github.com/cryptonotefoundation/cryptonote)
3. Achetez un ou deux serveurs dédiés basés sur Ubuntu (au moins 2 Go de RAM) pour les noeuds d'amorçage.


### Premier pas. Donnez un nom à votre pièce

** Un bon nom doit être unique. ** Vérifiez l'unicité avec [google] (http://google.com) et [Map of Coins] (mapofcoins.com) ou tout autre service similaire.

Le nom doit être spécifié deux fois:

**1. dans le fichier src/CryptoNoteConfig.h** - `CRYPTONOTE_NAME` constant

Exemple: 
```
const char CRYPTONOTE_NAME[] = "gallion";
```

**2. dans src/CMakeList.txt file** - set_property(TARGET daemon PROPERTY OUTPUT_NAME "YOURCOINNAME**d**")

Exemple: 
```
set_property(TARGET daemon PROPERTY OUTPUT_NAME "gallion")
```

** Note: ** Vous devez également modifier le nom du référentiel.


### Deuxième étape. Logique d'émission

**1. Masse monétaire totale ** (src/CryptoNoteConfig.h)

Quantité totale de pièces à émettre. La plupart des pièces à base de CryptoNote utilisent`(uint64_t)(-1)` (est égal à18446744073709551616). Vous pouvez définir un nombre explicitement (par exemple `UINT64_C(858986905600000000)`).

Exemple:
```
const uint64_t MONEY_SUPPLY = (uint64_t)(-1);
```

** 2. Courbe d'émission ** (src/CryptoNoteConfig.h)

Par défaut, CryptoNote fournit une formule d'émission avec une légère diminution de la récompense de bloc pour chaque bloc. Ceci est différent de Bitcoin où la récompense de bloc diminue de moitié tous les 4 ans.

La constante 'EMISSION_SPEED_FACTOR` définit la pente de la courbe d'émission. Ce paramètre est requis pour calculer la récompense de bloc.

Exemple:
```
const unsigned EMISSION_SPEED_FACTOR = 18;
```

** 3. Difficulté cible ** (src / CryptoNoteConfig.h)

La cible de difficulté est une période idéale entre les blocs. Dans le cas où un temps moyen entre blocs devient inférieur à la difficulté, la difficulté augmente. La cible de difficulté est mesurée en secondes.

La cible de difficulté influence directement plusieurs aspects du comportement de la pièce:

- vitesse de confirmation de la transaction: plus la durée entre les blocs est longue, plus la confirmation de la transaction est lente
- vitesse d'émission: plus le temps entre les blocs est long, plus le processus d'émission est lent
- taux d'orphelin: les chaînes avec des blocs très rapides ont un plus grand taux d'orphelin

Pour la plupart des pièces, la difficulté est de 60 ou 120 secondes.

Exemple:
```
const uint64_t DIFFICULTY_TARGET = 120;
```

**4. Formule de récompense de bloc**

Dans le cas où vous n'êtes pas satisfait de l'implémentation par défaut de la logique de récompense de bloc de CryptoNote, vous pouvez également la modifier. L'implémentation est dans`src/CryptoNoteCore/Currency.cpp`:
```
bool Currency::getBlockReward(size_t medianSize, size_t currentBlockSize, uint64_t alreadyGeneratedCoins, uint64_t fee, uint64_t& reward, int64_t& emissionChange) const
```

Cette fonction a deux parties:

- calcul de la récompense de base du bloc: `uint64_t baseReward = (m_moneySupply - alreadyGeneratedCoins) >> m_emissionSpeedFactor;`
- calcul de la pénalité de gros bloc: c'est ainsi que CryptoNote protège la chaîne de blocs contre les attaques par inondation de transaction et préserve les opportunités de croissance organique du réseau en même temps.

Seule la première partie de cette fonction est directement liée à la logique d'émission. Vous pouvez le changer comme vous le souhaitez. Voir MonetaVerde et DuckNote comme exemples où cette fonction est modifiée.


### Troisième étape. La mise en réseau

**1. Les ports par défaut pour les réseaux P2P et RPC** (src/CryptoNoteConfig.h)

Le port P2P est utilisé par les daemons pour se parler via le protocole P2P.
Le port RPC est utilisé par le portefeuille et d'autres programmes pour parler au daemons.

Il est préférable de choisir les ports qui ne sont pas utilisés par d'autres logiciels ou pièces de monnaie. Voir les listes de ports TCP connues:

* http://www.speedguide.net/ports.php
* http://www.networksorcery.com/enp/protocol/ip/ports00000.htm
* http://keir.net/portlist.html

Exemple:
```
const int P2P_DEFAULT_PORT = 17236;
const int RPC_DEFAULT_PORT = 18236;
```


**2.  Identifiant de réseau** (src/P2p/P2pNetworks.h)

Cet identifiant est utilisé dans les paquets réseau afin de ne pas mélanger deux réseaux cryptocoin différents. Changez tous les octets en valeurs aléatoires pour votre réseau:
```
const static boost::uuids::uuid CRYPTONOTE_NETWORK = { { 0xA1, 0x1A, 0xA1, 0x1A, 0xA1, 0x0A, 0xA1, 0x0A, 0xA0, 0x1A, 0xA0, 0x1A, 0xA0, 0x1A, 0xA1, 0x1A } };
```


**3. Noeud de départ** (src/CryptoNoteConfig.h)

Ajoutez des adresses IP de vos noeuds de départ.

Exemple:
```
const std::initializer_list<const char*> SEED_NODES = {
  "111.11.11.11:17236",
  "222.22.22.22:17236",
};
```


### Quatrième étape. Frais de transaction et paramètres associés

**1. Frais de transaction minimum** (src/CryptoNoteConfig.h)

Frais minimum zéro peut conduire à une inondation de transaction. Les transactions moins chères que les frais de transaction minimum ne seraient pas acceptées par les daemons. La valeur 100000 pour `MINIMUM_FEE` est généralement suffisante.

Exemple:
```
const uint64_t MINIMUM_FEE = 100000;
```


**2. Taille de bloc sans pénalité** (src/CryptoNoteConfig.h)

CryptoNote protège la chaîne contre l'inondation de tx en réduisant la récompense de bloc pour les blocs plus grands que la taille de bloc médiane. Cependant, cette règle s'applique aux blocs plus grands que les octets `CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE` bytes.

Exemple:
```
const size_t CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE = 20000;
```


### Cinquième étape. Adresse préfixe

Vous pouvez choisir une lettre (dans certains cas plusieurs lettres) toutes les adresses publiques de la pièce commenceront par. Il est défini par la constante `CRYPTONOTE_PUBLIC_ADDRESS_BASE58_PREFIX`. Puisque les règles pour les préfixes d'adresse sont non triviales, vous pouvez utiliser l'outil [générateur de préfixes] (https://cryptonotestarter.org/tools.html).

Example:
```
const uint64_t CRYPTONOTE_PUBLIC_ADDRESS_BASE58_PREFIX = 0xe9; // les adresses commencent par "f"
```


### Sixième étape. Bloc de la Genèse

**1. Construire les binaires avec genesis tx hex vide** (src/CryptoNoteConfig.h)

Vous devez laisser `const char GENESIS_COINBASE_TX_HEX []` vide et compiler les binaires sans lui.

Exemple:
```
const char GENESIS_COINBASE_TX_HEX[] = "";
```


** 2. Démarrez le démon pour imprimer le bloc de genèse **

Exécutez votre daemon avec l'argument `--print-genesis-tx`. Il imprimera le hachage de transaction coinbase du bloc genesis.

Exemple:
```
gallion --print-genesis-tx
```

**3. Copiez le hachage de transaction imprimé** (src/CryptoNoteConfig.h)

Copiez le hachage tx qui a été imprimé par le daemon dans `GENESIS_COINBASE_TX_HEX` in `src/CryptoNoteConfig.h`

Example:
```
const char GENESIS_COINBASE_TX_HEX[] = "013c01ff0001ffff...785a33d9ebdba68b0";
```


** 4. Recompilez les binaires **

Recompilez tout encore. Votre code de pièce est prêt maintenant. Faites une annonce pour les utilisateurs potentiels et amusez-vous!

## Construire CryptoNote 

### On *nix

Dépendances: GCC 4.7.3 ou ultérieur, CMake 2.8.6 ou ultérieur et Boost 1.55.

Vous pouvez les télécharger depuis:
* http://gcc.gnu.org/
* http://www.cmake.org/
* http://www.boost.org/
* Alternativement, il peut être possible de les installer en utilisant un gestionnaire de paquets.

Pour construire, placez-vous dans un répertoire où se trouve ce fichier et lancez `make`. Les exécutables résultants peuvent être trouvés dans `build/release/src`.

**Advanced options:**

* 
* Parallel build: lance `make -j<number of threads>` au lieu de `make`.
* Debug build: lance `make build-debug`.
* Suite de tests: exécutez `make test-release` pour exécuter des tests en plus de la construction. L'exécution de `make test-debug` fera de même avec la version de débogage.
* Construire avec Clang: il peut être possible d'utiliser Clang au lieu de GCC, mais cela peut ne pas fonctionner partout. Pour compiler, lancez `export CC=clang CXX=clang++` avant d'exécuter `make`.

### Sur Windows
Dépendances: MSVC 2013 ou ultérieur, CMake 2.8.6 ou ultérieur et Boost 1.55. Vous pouvez les télécharger depuis:

* http://www.microsoft.com/
* http://www.cmake.org/
* http://www.boost.org/

Pour construire, passez à un répertoire où se trouve ce fichier et exécutez les commandes: 
```
mkdir build
cd build
cmake -G "Visual Studio 12 Win64" ..
```

Et puis faites Build.
Bonne chance!# Gallion


ReleaseNotes.txt
Release notes Gallion 1.4.6

- Android platform support
- Added logging to blockchain synchronizer
- Refactored mnemonics
- Fixed memory leaks in mnemonics
- Fixed memory leak in simplewallet
- Fixed memory leak in cn_slow_hash
- Minor optimization of JSON KV-serializer
- LoggerMessage implementation is now platform-independent

Release notes Gallion 1.4.5

- Deterministic wallets and Mnemonic seed, courtesy of Monero Developers
- Various build fixes
- Transactions confirmations in RPC response for walletd
- Log file location for simplewallet
- New RPC methods
- Support for solo GPU mining

Release notes Gallion 1.4.4

- Transaction size limit
- Additional validation and attribution

Release notes Gallion 1.4.3

- Change of the Network ID
- Security update to address Disclosure of a Major Bug in CryptoNote Based Currencies by Monero developers

Release notes Gallion 1.4.2

- OpenAlias support in simplewallet
- More friendly simplewallet prompt

Release notes Gallion 1.4.1

- Additional transaction validation

Release notes Gallion 1.4.0

- GUI wallet mining support
- Simplewallet export/import keys
- Simplewallet import paperwallet/backup keys
- Simplewallet export/import tracking keys
- Change password in simplewallet



Release notes Gallion 1.3.0

- Bytecoin core 1.0.11 transition
- New difficulty algorithm
- Tail emission
- Daemon restricted RPC mode
- Fees for open remote node

Release notes 1.0.11

- New Bytecoin Wallet file format
- Daemon loading optimization

Release notes 1.0.10

- Blocksize increase
- Fusion transactions for Bytecoin RPC Wallet
- Bytecoin RPC Wallet API improvements

Release notes 1.0.9

- New API for Bytecoin RPC Wallet
- Various improvements

Release notes 1.0.8

- Fusion transactions for Bytecoin Wallet
- IWallet high-level API update
- JSON RPC improvements
- Synchronization improvements for OS X

Release notes 1.0.7

- Fusion transactions support
- Various simplewallet improvements

Release notes 1.0.6

- High-level API update
- Aggregate multi-addresses for Bytecoin RPC Wallet
- Wallet synchronization speed increase

Release notes 1.0.5

- High-level API for blockchain explorer
- Full network layer refactoring
- Transactions pool synchronization
- list_transactions method for RPC Wallet
- Various improvements

Release notes 1.0.4

- Bytecoin RPC Wallet
- New multithreading library
- Improved console logging
- Further optimizations

Release notes 1.0.3

- Multisignature API
- Low level API
- High level API improvements
- Instant transaction pool notifications
- Fully refactored simplewallet

Release notes 1.0.2

- Transaction history for simplewallet
- Reset command for simplewallet
- Various simplewallet improvements

Release notes 1.0.1

- Fix transfers in simplewallet

Release notes 1.0.0

- Multi-signatures
- Updated block reward scheme
- Further optimization in daemon RAM consumption
- Faster wallet refresh
- Transaction priority based on tx fee
- Transactions are returned from tx pools after 24 hours
- Dynamic maximum block size limit
- Reduced default transaction fee 
- Various network health updates

Release notes 0.8.11

- Increased minimum transaction fee
- Transaction pool optimizations
- High level API implementation
- CryptoNight hash function optimization
- Improvements for wallet JSON RPC API

Release notes 0.8.10

- Optimized blockchain storage memory usage
- Various code improvements

Release notes 0.8.9

- JSON RPC v2.0 compatibility
- JSON RPC over TCP added
- Further optimizations

Release notes 0.8.8

- Fixed a bug with checkpoints behavior
- UI improvements for daemon
- Various updates from our forks

Release notes 0.8.7

- OSX support
- Further optimizations

Release notes 0.8.6

- Simplwallet can set extra for transfers
- Improvements in JSON RPC for wallet
- UX improvements in simplewallet
- Win32 compilation

Release notes 0.8.5

- Port mapping for daemon using UPnP
- Improvements for daemon JSON RPC API

Release notes 0.8.4

- Formalization of high level API
- Improvements for wallet JSON RPC API

Release notes 0.8.3

- JSON RPC for wallet
- fixed bug with blockchain storing
- fixed bug with correct display of transfer's change 
- bug fix in simple wallet

Release notes 0.8.2

- now wallet is still working when daemon stores blockchain and can't serve clients; 
- no-console option for a daemon; 
- fixed problem with network synchronization; 
- major bug fix in simple wallet. 

Release notes 0.8.1

Bytecoin project is moved to GitHub

src/CryptoNoteConfig.h
#pragma once

#include <cstdint>

namespace CryptoNote {
	namespace parameters {

		const uint64_t CRYPTONOTE_MAX_BLOCK_NUMBER = 500000000;
		const size_t   CRYPTONOTE_MAX_BLOCK_BLOB_SIZE = 500000000;
		const size_t   CRYPTONOTE_MAX_TX_SIZE = 1000000000;
		const uint64_t CRYPTONOTE_PUBLIC_ADDRESS_BASE58_PREFIX = 0x65ed; // addresses start with "gmt"
		const size_t   CRYPTONOTE_MINED_MONEY_UNLOCK_WINDOW = 10;
		const uint64_t CRYPTONOTE_BLOCK_FUTURE_TIME_LIMIT = 60 * 60 * 2;

		const size_t   BLOCKCHAIN_TIMESTAMP_CHECK_WINDOW = 60;

		// MONEY_SUPPLY - total number coins to be generated
		const uint64_t MONEY_SUPPLY = UINT64_C(-1);
		const uint64_t TAIL_EMISSION_REWARD = UINT64_C(1000000000000);
		const size_t CRYPTONOTE_COIN_VERSION = 1;
		const unsigned EMISSION_SPEED_FACTOR = 22;
		static_assert(EMISSION_SPEED_FACTOR <= 8 * sizeof(uint64_t), "Bad EMISSION_SPEED_FACTOR");

		const size_t   CRYPTONOTE_REWARD_BLOCKS_WINDOW = 100;
		const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE = 1000000; //size of block (bytes) after which reward for block calculated using block size
		const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V2 = 1000000;
		const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V1 = 100000;
		const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_CURRENT = CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE;
		const size_t   CRYPTONOTE_COINBASE_BLOB_RESERVED_SIZE = 600;
		const size_t   CRYPTONOTE_DISPLAY_DECIMAL_POINT = 11;
		const uint64_t MINIMUM_FEE = UINT64_C(100000000);
		const uint64_t DEFAULT_DUST_THRESHOLD = MINIMUM_FEE;

		const uint64_t DIFFICULTY_TARGET = 240; // seconds
		const uint64_t EXPECTED_NUMBER_OF_BLOCKS_PER_DAY = 24 * 60 * 60 / DIFFICULTY_TARGET;
		const size_t   DIFFICULTY_WINDOW = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY; // blocks
		const size_t   DIFFICULTY_WINDOW_V2 = 17; // blocks
		const size_t   DIFFICULTY_CUT = 60;  // timestamps to cut after sorting
		const size_t   DIFFICULTY_LAG = 15;  // !!!
		static_assert(2 * DIFFICULTY_CUT <= DIFFICULTY_WINDOW - 2, "Bad DIFFICULTY_WINDOW or DIFFICULTY_CUT");

		const size_t   MAX_BLOCK_SIZE_INITIAL = 1000000;
		const uint64_t MAX_BLOCK_SIZE_GROWTH_SPEED_NUMERATOR = 100 * 1024;
		const uint64_t MAX_BLOCK_SIZE_GROWTH_SPEED_DENOMINATOR = 365 * 24 * 60 * 60 / DIFFICULTY_TARGET;

		const uint64_t CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_BLOCKS = 1;
		const uint64_t CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_SECONDS = DIFFICULTY_TARGET * CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_BLOCKS;

		const uint64_t CRYPTONOTE_MEMPOOL_TX_LIVETIME = 60 * 60 * 24;     //seconds, one day
		const uint64_t CRYPTONOTE_MEMPOOL_TX_FROM_ALT_BLOCK_LIVETIME = 60 * 60 * 24 * 7; //seconds, one week
		const uint64_t CRYPTONOTE_NUMBER_OF_PERIODS_TO_FORGET_TX_DELETED_FROM_POOL = 7;  // CRYPTONOTE_NUMBER_OF_PERIODS_TO_FORGET_TX_DELETED_FROM_POOL * CRYPTONOTE_MEMPOOL_TX_LIVETIME = time to forget tx

		const size_t   FUSION_TX_MAX_SIZE = CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_CURRENT * 30 / 100;
		const size_t   FUSION_TX_MIN_INPUT_COUNT = 12;
		const size_t   FUSION_TX_MIN_IN_OUT_COUNT_RATIO = 4;

		const uint32_t UPGRADE_HEIGHT_V2 = 60000;
		const uint32_t UPGRADE_HEIGHT_V3 = 4294967294;
		const unsigned UPGRADE_VOTING_THRESHOLD = 90;               // percent
		const uint32_t   UPGRADE_VOTING_WINDOW = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY;  // blocks
		const uint32_t   UPGRADE_WINDOW = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY;  // blocks
		static_assert(0 < UPGRADE_VOTING_THRESHOLD && UPGRADE_VOTING_THRESHOLD <= 100, "Bad UPGRADE_VOTING_THRESHOLD");
		static_assert(UPGRADE_VOTING_WINDOW > 1, "Bad UPGRADE_VOTING_WINDOW");

		const char     CRYPTONOTE_BLOCKS_FILENAME[] = "blocks.dat";
		const char     CRYPTONOTE_BLOCKINDEXES_FILENAME[] = "blockindexes.dat";
		const char     CRYPTONOTE_BLOCKSCACHE_FILENAME[] = "blockscache.dat";
		const char     CRYPTONOTE_POOLDATA_FILENAME[] = "poolstate.bin";
		const char     P2P_NET_DATA_FILENAME[] = "p2pstate.bin";
		const char     CRYPTONOTE_BLOCKCHAIN_INDICES_FILENAME[] = "blockchainindices.dat";
		const char     MINER_CONFIG_FILE_NAME[] = "miner_conf.json";
	} // parameters

	const char     CRYPTONOTE_NAME[] = "gallion";
	const char     GENESIS_COINBASE_TX_HEX[] = "010a01ff0001ffffffffff7f029b2e4c0281c0b02e7c53291a94d1d0cbff8883f8024f5142ee494ffbbd08807121015cf9e7b15a00a9a8bdda276d5e3c439b75b64cd9022bb39f412e3faef2b91f0e";

	const uint8_t  CURRENT_TRANSACTION_VERSION = 1;
	const uint8_t  BLOCK_MAJOR_VERSION_1 = 1;
	const uint8_t  BLOCK_MAJOR_VERSION_2 = 2;
	const uint8_t  BLOCK_MAJOR_VERSION_3 = 3;
	const uint8_t  BLOCK_MINOR_VERSION_0 = 0;
	const uint8_t  BLOCK_MINOR_VERSION_1 = 1;

	const size_t   BLOCKS_IDS_SYNCHRONIZING_DEFAULT_COUNT = 10000;  //by default, blocks ids count in synchronizing
	const size_t   BLOCKS_SYNCHRONIZING_DEFAULT_COUNT = 200;    //by default, blocks count in blocks downloading
	const size_t   COMMAND_RPC_GET_BLOCKS_FAST_MAX_COUNT = 1000;

	const int      P2P_DEFAULT_PORT = 6012;
	const int      RPC_DEFAULT_PORT = 1119;

	const size_t   P2P_LOCAL_WHITE_PEERLIST_LIMIT = 5050;
	const size_t   P2P_LOCAL_GRAY_PEERLIST_LIMIT = 5000;

	const size_t   P2P_CONNECTION_MAX_WRITE_BUFFER_SIZE = 64 * 1024 * 1024; // 64 MB
	const uint32_t P2P_DEFAULT_CONNECTIONS_COUNT = 8;
	const size_t   P2P_DEFAULT_WHITELIST_CONNECTIONS_PERCENT = 70;
	const uint32_t P2P_DEFAULT_HANDSHAKE_INTERVAL = 60;            // seconds
	const uint32_t P2P_DEFAULT_PACKET_MAX_SIZE = 50000000;      // 50000000 bytes maximum packet size
	const uint32_t P2P_DEFAULT_PEERS_IN_HANDSHAKE = 250;
	const uint32_t P2P_DEFAULT_CONNECTION_TIMEOUT = 5000;          // 5 seconds
	const uint32_t P2P_DEFAULT_PING_CONNECTION_TIMEOUT = 2000;          // 2 seconds
	const uint64_t P2P_DEFAULT_INVOKE_TIMEOUT = 60 * 2 * 1000; // 2 minutes
	const size_t   P2P_DEFAULT_HANDSHAKE_INVOKE_TIMEOUT = 5000;          // 5 seconds
	const char     P2P_STAT_TRUSTED_PUB_KEY[] = "8f80f9a5a434a9f1510d13336228debfee9c918ce505efe225d8c94d045fa115";

	const char* const SEED_NODES[] = {
		"gallion.ddns.net:6012",
		"gallion.ddns.net:6013"
	};

	struct CheckpointData {
		uint32_t height;
		const char* blockId;
	};

	const std::initializer_list<CheckpointData> CHECKPOINTS = {
		{2000,	"915357c2512935fc334d19475c46dbe876c420c18f05bb4cf771fc2b3ca34d6c" }
	};

} // CryptoNote

#define ALLOW_DEBUG_COMMANDS


